### ISSUE TITLE
Using Spock with KafkaContainer it fails 

### ISSUE THREAD

**Issue by umair-akram** [testcontainers/testcontainers-java#1106]
> @DirtiesContext
> @SpringBootTest
> @ContextConfiguration(initializers = Intializer )
> abstract class BaseIntegrationTest extends Specification {
> public static KafkaContainer kafka = new KafkaContainer()
> static {
>     BaseIntegrationTest.kafka.start()
> }
> static  class Intializer implements ApplicationContextInitializer<ConfigurableApplicationContext>{
>     @Override
>     void initialize(ConfigurableApplicationContext applicationContext) {
>         System.setProperty("kafka.consumer.endpoint", BaseIntegrationTest.kafka.bootstrapServers.replace("PLAINTEXT://", ""))
>     }
> }
> }

I have posted my code above I have a lot of tests that inherit this class. When I run any test class from **Intellij** it works fine but On circle CI the build fails with the following error.

`at org.spockframework.runtime.BaseSpecRunner.createSpecInstance(BaseSpecRunner.java:113)
	at org.spockframework.runtime.BaseSpecRunner.run(BaseSpecRunner.java:65)
	at org.spockframework.runtime.Sputnik.run(Sputnik.java:63)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:116)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:59)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:39)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy1.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:109)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:146)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:128)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ExceptionInInitializerError
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.spockframework.runtime.BaseSpecRunner.createSpecInstance(BaseSpecRunner.java:107)
	... 32 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Container startup failed
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:199)
	at org.testcontainers.containers.KafkaContainer.start(KafkaContainer.java:67)
	at org.testcontainers.containers.KafkaContainer$start.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:110)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:114)
	... 38 more
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:83)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:192)
	... 44 more
Caused by: java.lang.NullPointerException: containerId was not specified
	at org.testcontainers.shaded.com.google.common.base.Preconditions.checkNotNull(Preconditions.java:228)
	at com.github.dockerjava.core.command.LogContainerCmdImpl.withContainerId(LogContainerCmdImpl.java:78)
	at com.github.dockerjava.core.command.LogContainerCmdImpl.<init>(LogContainerCmdImpl.java:38)
	at com.github.dockerjava.core.DockerClientImpl.logContainerCmd(DockerClientImpl.java:335)
	at org.testcontainers.dockerclient.AuditLoggingDockerClient.logContainerCmd(AuditLoggingDockerClient.java:22)
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:255)
	at org.testcontainers.containers.GenericContainer.lambda$start$0(GenericContainer.java:194)
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:76)`

**Comment by bsideup** [comment]
Please check my answer:
https://github.com/testcontainers/testcontainers-java/issues/791#issuecomment-450347133

I'm closing this one as a Duplicate of #791

**Comment by umair-akram** [comment]
I just checked that link I am not sure what I am missing here.

**Comment by bsideup** [comment]
@umair-akram 
```
version: 2
executorType: machine
```

**Comment by umair-akram** [comment]
```yml
version: 2
executorType: machine
references:
  container_config: &container_config
    executorType: machine
    machine:
      docker_layer_caching: true
  restore_gradle_cache: &restore_gradle_cache
    restore_cache:
      keys:
        - v1-gradle-{{ checksum ".checksum" }}
        - v1-gradle-
  restore_git_cache: &restore_git_cache
    restore_cache:
      keys:
        - v1-git-{{ .Branch }}-{{ .Revision }}
        - v1-git-{{ .Branch }}-
        - v1-git-
  save_gradle_cache: &save_gradle_cache
    save_cache:
      key: v1-gradle-{{ checksum ".checksum" }}
      paths:
        - ~/.gradle
        - ~/.m2
  save_git_cache: &save_git_cache
    save_cache:
      key: v1-git-{{ .Branch }}-{{ .Revision }}
      paths:
        - .git
  checksum: &checksum
    run:
      name: Get checksum of all Gradle files
      command: md5sum {,**/}*.gradle > .checksum
  gather_test_results: &gather_test_results
    run:
      name: Gather test results
      shell: /bin/bash
      command: |
        mkdir -p ~/junit/
        find . -type f -regex ".*/build/test-results/.*xml" -exec cp {} ~/junit/ \;
        pushd service/build/reports/tests/test && zip -r ~/junit_html.zip * && popd || true
      when: always
jobs:
  build:
    <<: *container_config
    resource_class: medium+
    environment:
      _JAVA_OPTIONS: "-Xmx2048m"
      GRADLE_OPTS: '-Dorg.gradle.console=plain -Dorg.gradle.parallel=true'
    steps:
      - *restore_git_cache
      - checkout
      - *checksum
      - *restore_gradle_cache
      - run:
          name: check docker version
          command: |
            docker -v
      - run:
          name: Run tests
          command: ./gradlew clean check --no-daemon --stacktrace
      - *gather_test_results
      - store_test_results:
          path: ~/junit
      - store_artifacts:
          path: ~/junit-html.zip
      - *save_gradle_cache
      - *save_git_cache
```
Whats wrong with file still getting the very same error if I remove parallel execution from the gradle command tests start to work.

**Comment by bsideup** [comment]
> resource_class: medium+

Maybe not enough resources?

**Comment by umair-akram** [comment]
```java
String networkAlias = getNetworkAliases().get(0);
        proxy = new SocatContainer()
                .withNetwork(getNetwork())
                .withTarget(9092, networkAlias)
                .withTarget(2181, networkAlias);
```

That's what I found in the `start` method of kafka container `9092, 2181` as exposed ports can they cause any troubles with `-Dorg.gradle.parallel=true`

**Comment by bsideup** [comment]
@umair-akram when you run with `parallel=true`, you start multiple Kafkas in parallel. 
Kafka is a very hungry system and consumes a lot of resources, and it fails to start when you run multiple (2? 3? 5?) kafkas in parallel.

**Comment by umair-akram** [comment]
```gradle
test {
	maxParallelForks = 1
	maxHeapSize = "2048m"
	jvmArgs = ['-XX:+TieredCompilation', '-XX:TieredStopAtLevel=1', '-Xverify:none']
}
```

**Comment by bsideup** [comment]
without `parallel=true`, Gradle will run the tests in the same JVM. With it, it will fork and allocate 2Gb heap (according to the snippet you posted). I guess CircleCI does not have *that* much memory allocated per CI executor

**Comment by umair-akram** [comment]
Hmmm .. that could be the case and I am trying it with the less memory size, but before I introduced this `test-containers` to the build process, everything worked fine with the same memory arguments.

**Comment by bsideup** [comment]
I guess you were using EmbeddedKafka? If so, it runs in the same JVM process and there is no additional memory allocation, but now you start a separate Kafka process (with Docker).
But, if I am right, you can easily lower heap size for your tests because previously it was shared with embedded Kafka, and your app was not using all 2Gb if heap

**Comment by umair-akram** [comment]
Yes I have lowered the heap size previously we were not using kafka I just introduced it to our service and heard of test-containers wanted to introduced them slowly to get rid of this embedded things from our tests. SoI picked `kafka-container`  to be the first then also gonna use it for `Postgress` as well.

**Comment by kiview** [comment]
Although the issue is closed, I've edited the title to reflect the actual problem (nothing Spock related here).
